Update app-go to use 5.0 Drivers (#33)
neo4j-graphacademy/courses: asciidoc/courses/app-go/modules/1-driver/lessons/1-installation/lesson.adoc

= Installation
// :type: video
:type: lesson
:branch: main

In the link:/courses/cypher-fundamentals/[Cypher Fundamentals^] course, we cover how to query Neo4j using a language called Cypher.
To execute a Cypher statement against a Neo4j database you will use an object called a *Driver*.

To connect to and query Neo4j from within a Go application, you use the link:https://neo4j.com/developer/go[Neo4j Go Driver^].

== Installing the Driver

The Neo4j Go Driver can be installed using the `go get` command.

include::{repository-raw}/{branch}/README.adoc[tag=install]

== Importing the Library

The `github.com/neo4j/neo4j-go-driver` package exports a `neo4j` object which contains several helper functions for interacting with Neo4j.

.Importing The Driver
[source,go,indent=0,subs="attributes+"]
----
include::{repository-raw}/{branch}/pkg/example/example.go[tag=import]
----

== Creating a Driver Instance

> The most prominent helper function on the `neo4j` object is the `.NewDriverWithContext()` function, which accepts neo4j credentials and returns a new `Driver` object.

Reviewer: the type is actually `DriverWithContext` since I could not remove the legacy context-less `Driver` type just yet

[TIP]
.Authentication Types
====
Neo4j supports four authentication methods, basic authentication (e.g. username and password or LDAP), a base64 encoded Bearer token, a Kerberos token, or a custom authentication token.
For more information on these options, visit link:https://neo4j.com/docs/operations-manual/current/authentication-authorization/[Authentication and Authorization^].
====

=== Verifying Connectivity

You can verify that the connection details used during driver instantiation are correct by calling the `VerifyConnectivity()` method.

This function returns a Promise which throws a `Neo.ClientError.Security.Unauthorized` error if the provided credentials are incorrect.

.Verify Connectivity
[source,go,indent=0,role=nocopy,subs="attributes+"]
----
include::{repository-raw}/{branch}/pkg/example/example.go[tag=verify]
----

== Sessions

From the Driver object, you will open one or more **sessions**.
A session is a client-side abstraction that borrows underlying TCP connections from a _connection pool_ held by the Driver and uses them to communicate with Neo4j.

Sessions should be considered lightweight and disposable and can be opened and closed at will.

> You should `defer` the closing of each session  object using the `PanicOnClosureError()` function, which will release any underlying TCP connections held by the session object once the surrounding context has completed.

Reviewer: note: `PanicOnClosureError` is just a convenience function used for the example (and it's quite crude, it panics whenever there is an error)
the more generic advice would be to call `Close` on sessions (and driver), preferably with a context that does not have a deadline (since you don't want to over-constrain the execution time required to close sessions/driver

> This practise should also be followed when you are finished using the driver instance.

Reviewer: ```suggestion
This practice should also be followed when you are finished using the driver instance.
```

.Session Lifecycle
[source,go]
----
include::{repository-raw}/{branch}/pkg/example/example.go[tags=driver,close,indent=0]

// Open a new Session
session := driver.NewSession(ctx, neo4j.SessionConfig{})
defer PanicOnClosureError(ctx, session)
----

== Running a Sample Query

To run a one-off Cypher query, you can use the `session.Run()` method.

.Running a One-off Query
[source,go,indent=0,role=nocopy,subs="attributes+"]
----
include::{repository-raw}/{branch}/pkg/example/example.go[tag=oneoff]
----

The method accepts four arguments:

1. The execution context.
2. A string to represent a Cypher statement

> 3. Optionally, a map representing query parameters (prefixed in Cypher with `$`)

Reviewer: ```suggestion
3. a map representing query parameters (can be `nil`, prefixed in Cypher with `$`)
```
it is sadly not optional

> 4. Also optional, a map containing transaction config.

Reviewer: ```suggestion
4. Optional, a map containing transaction config.
```

[WARNING]
.For one-off queries only

> If there are any transient errors when running a query, the Driver will not attempt to retry a query when using `session.Run()`.

Reviewer: ```suggestion
If there are any transient errors when running a query, the driver will not attempt to retry a query when using `session.Run()`.
```

> For this reason, these should only be used for one-off queries and not in production.

Reviewer: not sure if `CALL {} IN TRANSACTIONS` and `... USING PERIODIC COMMIT` are worth mentioning here? (they HAVE to run with `session.Run`, other execution methods do not work)

Author: That's interesting. I didn't know that.  I'm not sure whether this is out of the scope of the course

.Accessing Results
[source,go,indent=0,role=nocopy,subs="attributes+"]
----
include::{repository-raw}/{branch}/pkg/example/example.go[tag=oneoffresult]
----

In the code sample above, the `p` value is extracted from each node using the `neo4j.GetRecordValue()` function, which is assigned to the `person` variable.
In this instance, `p` refers to a Neo4j Node, which is is hydrated into an instance of the `neo4j.Node` interface.

Should the `p` key not exist on the record, or any other error occurs while trying to retrieve the value, the subsequent error will be assigned to the `err` variable.

We will cover result processing further in link:../3-execute[Lesson 3: Read and Write Transactions^].

// == Closing the Driver

// The Driver provides a `.close()` method.
// Calling this function will close any open sessions, disconnect all servers and release any resources held by the driver.

// .Running a One-off Query
// [source,go,indent=0,role=nocopy,subs="attributes+"]
// ----
// include::{repository-raw}/{branch}/pkg/example/example.go[tag=close]
// ----

== Check Your Understanding

include::questions/1-package-name.adoc[leveloffset=+1]

[.summary]
== Summary

> In this lesson, you learned the basics of the Neo4j Go Driver, how to install the dependency, create a new Driver instance and run a sample query.

Reviewer: ```suggestion
In this lesson, you learned the basics of the Neo4j Go Driver, how to install the dependency, create a new driver instance and run a sample query.
```
